extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,287 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpAdUIHandler} from '../../amp-ad/0.1/amp-ad-ui'; // eslint-disable-line no-unused-vars
+import {
+  LayoutInfoDef,
+  getAmpAdMetadata,
+  sendXhrRequest,
+} from '../../amp-a4a/0.1/a4a-utils';
+import {
+  RendererDef,
+  RendererInputDef,
+  ValidatorDef, // eslint-disable-line no-unused-vars
+  ValidatorOutputDef,
+  ValidatorResultType, // eslint-disable-line no-unused-vars
+} from '../../amp-a4a/0.1/a4a-render';
+import {dev} from '../../../src/log';
+import {utf8Encode} from '../../../src/utils/bytes';
+
+const TAG = 'amp-ad-network-base';
+
+export class AmpAdNetworkBase extends AMP.BaseElement {
+
+  constructor(element) {
+    super(element);
+
+    /** @private {Object<ValidatorResultType, !RendererDef>} */
+    this.boundRenderers_ = {};
+
+    /** @private {?ValidatorDef} */
+    this.boundValidator_ = null;
+
+    /** @private {?string} */
+    this.adUrl_ = null;
+
+    /** @private {?string} */
+    this.expandedAdUrl_ = null;
+
+    /** @private {?ArrayBuffer} */
+    this.unvalidatedBytes_ = null;
+
+    /** @private {!LayoutInfoDef} */
+    this.initialSize_ = {
+      // TODO(levitzky) handle non-numeric values.
+      width: element.getAttribute('width'),
+      height: element.getAttribute('height'),
+    };
+
+    /** @private {string} @const */
+    this.networkType_ = element.getAttribute('type') || 'anon';
+
+    /**
+     * @private {number} unique ID of the currently executing promise to allow
+     * for cancellation.
+     */
+    this.freshnessId_ = 0;
```

Comment:
what is this?

Reply:
This serves the same purpose as [promiseId_ in amp-a4a](https://github.com/ampproject/amphtml/blob/master/extensions/amp-a4a/0.1/amp-a4a.js#L258). The name was changed due to the fact that we don't have a single promise chain like we do in amp-a4a.